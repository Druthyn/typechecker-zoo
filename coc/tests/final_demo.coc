inductive Nat : Type with
  | zero : Nat
  | succ : Nat -> Nat

inductive Bool : Type with
  | true : Bool
  | false : Bool

axiom mult : Nat -> Nat -> Nat

def compose (A : Type) (B : Type) (C : Type) (g : B -> C) (f : A -> B) (x : A) : C :=
  g (f x)

def square (x : Nat) : Nat :=
  mult x x

def doTwice (A : Type) (h : A -> A) (x : A) : A :=
  h (h x)

def doThrice (A : Type) (h : A -> A) (x : A) : A :=
  h (h (h x))