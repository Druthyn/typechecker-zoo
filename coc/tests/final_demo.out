Parsed module with 7 declarations
âœ“ Module 'tests/final_demo.coc' type checks successfully!